### PR TITLE
Port to guile 3.

### PR DESCRIPTION
**Pull request by marxin**
This is port to Guile 3 version as mentioned in:
https://bugs.gnucash.org/show_bug.cgi?id=797579

However, I still face some failures:
```
$ ctest --VV
...
 79/111 Test  #79: test-standard-category-report ....***Failed    0.16 sec
        Start  80: test-standard-net-linechart
 80/111 Test  #80: test-standard-net-linechart ......***Failed    0.14 sec
        Start  81: test-standard-net-barchart
 81/111 Test  #81: test-standard-net-barchart .......***Failed    0.14 sec
        Start  82: test-cashflow-barchart
 82/111 Test  #82: test-cashflow-barchart ...........***Failed    0.14 sec
        Start  83: test-charts
 83/111 Test  #83: test-charts ......................***Failed    0.37 sec
        Start  84: test-transaction
 84/111 Test  #84: test-transaction .................***Failed    0.30 sec
        Start  85: test-account-summary
 85/111 Test  #85: test-account-summary .............***Failed    0.15 sec
        Start  86: test-balsheet-pnl
 86/111 Test  #86: test-balsheet-pnl ................***Failed    0.20 sec
        Start  87: test-income-gst
 87/111 Test  #87: test-income-gst ..................***Failed    0.15 sec
        Start  88: test-budget
 88/111 Test  #88: test-budget ......................***Failed    0.20 sec
        Start  89: test-register
 89/111 Test  #89: test-register ....................***Failed    0.18 sec
        Start  90: test-trial-balance
 90/111 Test  #90: test-trial-balance ...............***Failed    0.17 sec
        Start  91: test-equity-statement
 91/111 Test  #91: test-equity-statement ............***Failed    0.15 sec
        Start  92: test-average-balance
 92/111 Test  #92: test-average-balance .............***Failed    0.15 sec
        Start  93: test-invoice
 93/111 Test  #93: test-invoice .....................***Failed    0.18 sec
        Start  94: test-new-owner-report
 94/111 Test  #94: test-new-owner-report ............***Failed    0.19 sec
        Start  95: test-owner-report
 95/111 Test  #95: test-owner-report ................***Failed    0.17 sec
        Start  96: test-portfolios
 96/111 Test  #96: test-portfolios ..................***Failed    0.14 sec
        Start  97: test-stress-options
```

Any hint how can I debug the failures?

### CODE REVIEW

**Comment by marxin** [comment]
> Any hint how can I debug the failures?

It's fine on Travis CI, so that I'm removing WIP keyword.

File: po/gnucash-pot.cmake [hunk 1]
```diff
@@ -20,7 +20,7 @@ execute_process(
                         --flag=g_snprintf:3:c-format
                         --flag=g_scanner_error:2:c-format
                         --flag=g_scanner_warn:2:c-format
-                        --keyword=_
+                        --keyword=G_
```
Comment (gjanssens) [inline po/gnucash-pot.cmake:23]: You can't remove the `_` keyword. It's still in use by C and C++ code. Just add the `G_` keyword instead.

File: bindings/guile/core-utils.scm [hunk 2]
```diff
@@ -38,8 +38,8 @@
 
 (define-public gnc:version (gnc-version))
 ;; gettext functions
-(define-public _ gnc:gettext)
+(define-public G_ gnc:gettext)
 (define-syntax N_
     (syntax-rules ()
-        ((_ x) x)))
+        ((G_ x) x)))
```
Comment (christopherlam) [inline bindings/guile/core-utils.scm:44]: No; in this `((_ x) x)))`, the `_` is part of macro syntax. It must be left alone. A more concise definition of `N_` is:
```scheme
(define-syntax-rule (N_ x) x)
```

**Comment by marxin** [comment]
> Looks ok to me now.

Thanks.

> The contents of gnucash.pot is identical before and after applying this PR.
> And running this on an installation with guile 2.2 message translations for guile code are found properly. I can't test with guile 3 atm as I don't have an installation with that version.

You can easily test it with `opensuse/tumbleweed` Docker image where we have guile 3.0. But you may miss devel packages that needs to be installed on `openSUSE`.

> 
> I'm inclined to apply it for 4.0 still (it won't break string freeze), but this may have to be discussed first.

**Comment by christopherlam** [comment]
Some experiments: I use ubuntu.
- guile-2.2-dev installed
- `sudo apt install guile-3.0-dev`
- `ninja` - compiles completely. lots of deprecation warnings.
- launch binary: fails - tries to load .go from 2.2 cache

**Comment by christopherlam** [comment]
Latest experiment.

guix environment --ad-hoc guile
cmake etc
ninja

```
[408/994] Generating ../../lib/guile/3.0/site-ccache/tests/unittest-support.go
FAILED: lib/guile/3.0/site-ccache/tests/unittest-support.go 
cd /home/chris/sources/gnucash/master3-build/common/test-core && /usr/bin/cmake -E env LD_LIBRARY_PATH=/home/chris/sources/gnucash/master3-build/lib:/home/chris/sources/gnucash/master3-build/lib/gnucash: GNC_UNINSTALLED=YES GNC_BUILDDIR=/home/chris/sources/gnucash/master3-build GUILE_LOAD_PATH=/home/chris/sources/gnucash/master-3/common/test-core:/home/chris/sources/gnucash/master3-build/common/test-core:/home/chris/sources/gnucash/master3-build/common/test-core/deprecated:/home/chris/sources/gnucash/master3-build/share/guile/site/3.0 GUILE_LOAD_COMPILED_PATH=/home/chris/sources/gnucash/master3-build/common/test-core:/home/chris/sources/gnucash/master3-build/lib/guile/3.0/site-ccache:/home/chris/sources/gnucash/master3-build/lib/guile/3.0/site-ccache/gnucash/deprecated GNC_MODULE_PATH=/home/chris/sources/gnucash/master3-build/lib:/home/chris/sources/gnucash/master3-build/lib/gnucash: /gnu/store/gwc3q4asf4y4bsvabkwj37mnls9fn2mv-profile/bin/guile -e "(@@ (guild) main)" -s /gnu/store/gwc3q4asf4y4bsvabkwj37mnls9fn2mv-profile/bin/guild compile -o /home/chris/sources/gnucash/master3-build/lib/guile/3.0/site-ccache/tests/unittest-support.go /home/chris/sources/gnucash/master-3/common/test-core/unittest-support.scm
Backtrace:
In ice-9/boot-9.scm:
    718:2 19 (call-with-prompt _ _ #<procedure default-prompt-handle…>)
In ice-9/eval.scm:
    619:8 18 (_ #(#(#<directory (guile-user) 7fd902c84f00>)))
In /gnu/store/9zw374g3xkm59hgf5ig9057ll0iv7qjj-profile/bin/guild:
    72:17 17 (main _)
In srfi/srfi-1.scm:
    634:9 16 (for-each #<procedure 7fd903096380 at scripts/compile.…> …)
In scripts/compile.scm:
   263:26 15 (_ _)
In system/base/target.scm:
     65:6 14 (with-target _ _)
In system/base/compile.scm:
    152:6 13 (compile-file _ #:output-file _ #:from _ #:to _ #:env _ …)
     43:4 12 (call-with-output-file/atomic _ _ _)
In ice-9/boot-9.scm:
  1736:10 11 (with-exception-handler _ _ #:unwind? _ # _)
In system/base/compile.scm:
    59:11 10 (_)
   155:11  9 (_ #<closed: file 7fd901d9e070>)
   235:18  8 (read-and-compile #<input: unittest-support.scm 15> # _ …)
   183:32  7 (compile-fold (#<procedure compile-tree-il (x e opts)>) …)
In ice-9/boot-9.scm:
   2806:4  6 (save-module-excursion #<procedure 7fd9018889a0 at lang…>)
In language/scheme/compile-tree-il.scm:
    31:15  5 (_)
In ice-9/psyntax.scm:
  1241:36  4 (expand-top-sequence ((eval-when (compile load # #) #)) …)
  1233:19  3 (parse _ (("placeholder" placeholder)) ((top) #(# # …)) …)
   285:10  2 (parse _ (("placeholder" placeholder)) ((top) #(# # …)) …)
In unknown file:
           1 (load-extension "libtest-core-guile" "scm_init_unittest…")
In ice-9/boot-9.scm:
  1669:16  0 (raise-exception _ #:continuable? _)

ice-9/boot-9.scm:1669:16: In procedure raise-exception:
In procedure dynamic-link: file: "libtest-core-guile", message: "file not found"
[413/994] Generating gnucash.desktop.in
ninja: build stopped: subcommand failed.
```

**Comment by jralls** [comment]
Is lib/libtest-core-guile.so present? Was it built with the right libguile.so?

**Comment by marxin** [comment]
In my case, I see:
```
$ ldd lib64/gnucash/libtest-core-guile.so | grep guile
	libguile-3.0.so.1 => /usr/lib64/libguile-3.0.so.1 (0x00007ffff7e3e000)
```

**Comment by jralls** [comment]
@marxin and in your case does `guild` fail to build `unittest-support.scm` as it does for @christopherlam?

**Comment by marxin** [comment]
There's my build log:
[gnucash.txt](https://github.com/Gnucash/gnucash/files/4823531/gnucash.txt)

**Comment by gjanssens** [comment]
> There's my build log:
> [gnucash.txt](https://github.com/Gnucash/gnucash/files/4823531/gnucash.txt)

Can you export
```
CTEST_OUTPUT_ON_FAILURE=On
```
in your shell before running `make check` ? The resulting output will give more details on the failed tests.

So for example, please run

```
make
CTEST_OUTPUT_ON_FAILURE=On make check
```

**Comment by marxin** [comment]
Here it is:
[gnucash-2.txt](https://github.com/Gnucash/gnucash/files/4824534/gnucash-2.txt)

**Comment by gjanssens** [comment]
@marxin Thanks. So all failing tests seem to be caused by the same error: a parse error which has to do with XMLNS. The trace files suggest gnc:options->sxml doesn't work as it should with guile 3. I applied the patch on my Fedora 31 box (using guile 2.2 still) and that works ok. So it's not that the patch broke something, it's that this particular function still needs tweaks for guile 3. Either guile 3 renders our reports differently or xml->sxml's behaviour has changed.

It's also a different error from what @christopherlam reports in guix. I don't know what happens there.

I'm in the process of setting up a Ubuntu 20.04 VM to see if I can help out.

**Comment by gjanssens** [comment]
Meanwhile @marxin can you attach the html of anygenerated report ? You can for example run gnucash, run a report and with that all open search /tmp for files whose names start with gnc-report-...
Please provide a link to one of those as well. I'd be curious to see if the generated html is different with guile 3. (If you can run reports with gnucash/guile 3 of course).

**Comment by marxin** [comment]
There's a generate report:
[yyy.html.txt](https://github.com/Gnucash/gnucash/files/4825674/yyy.html.txt)

**Comment by christopherlam** [comment]
The report you attached suggests that the stylesheet was not loaded properly. Sounds like a module dependency issue.

**Comment by marxin** [comment]
Sorry, I have the project built remotely and I can't run it with `ssh -X ..` so I tried to copy all the content of `objdir` and I run it locally.

**Comment by gjanssens** [comment]
The build in my Ubuntu 20.04 VM with guile 3 gives the same test errors. The program itself runs, but even there I see some glitches, like the currency formatting on [this report](https://imgur.com/a/FnSyGCZ).

**Comment by christopherlam** [comment]
Geert this means the stylesheets were not loaded. They are responsible for formatting monetary objects into strings.

**Comment by gjanssens** [comment]
Interesting. That changes the question to why aren't they loaded ?

**Comment by gjanssens** [comment]
@christopherlam It's clearly more subtle than that. I can select the stylesheets in the report and the report responds to changing the stylesheet: [the same report with technicolor stylesheet](https://imgur.com/a/KqtGBfD).

**Comment by gjanssens** [comment]
From what I can see I believe the test failures are due to the same issue. The poorly rendered numbers in guile 3 seem to choke xml->sxml. I suspect this to be because there's no dtd description for nodes like this:
```
<gnc-monetary> #<<gnc-monetary> commodity: #<swig-pointer gnc_commodity * 558de280b400> amount: 3>
```

**Comment by gjanssens** [comment]
As per @christopherlam 's [request on irc](https://lists.gnucash.org/logs/2020/06/24.html#T22:44:57) I have added a debug statement in https://github.com/Gnucash/gnucash/blob/master/gnucash/report/html-document.scm#L309

This is the output when running test-invoice afterwards (which is one of the failing tests):
```
$ CTEST_OUTPUT_ON_FAILURE=on ctest -V -R 'test-invoice'
UpdateCTestConfiguration  from :/home/janssege/Development/gnucash/builds/master/DartConfiguration.tcl
UpdateCTestConfiguration  from :/home/janssege/Development/gnucash/builds/master/DartConfiguration.tcl
Test project /home/janssege/Development/gnucash/builds/master
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 93
    Start 93: test-invoice

93: Test command: /usr/bin/cmake "-E" "env" "/usr/bin/guile" "--debug" "-c" "(load-from-path "test-invoice")(exit (run-test))"
93: Environment variables: 
93:  GNC_UNINSTALLED=yes
93:  GNC_BUILDDIR=/home/janssege/Development/gnucash/builds/master
93:  LD_LIBRARY_PATH=/home/janssege/Development/gnucash/builds/master/lib:/home/janssege/Development/gnucash/builds/master/lib/gnucash:
93:  GNC_MODULE_PATH=/home/janssege/Development/gnucash/builds/master/lib:/home/janssege/Development/gnucash/builds/master/lib/gnucash
93:  GUILE=/usr/bin/guile
93:  GUILE_LOAD_PATH=/home/janssege/Development/gnucash/builds/master/share/guile/site/3.0/gnucash/deprecated
93:  GUILE_LOAD_COMPILED_PATH=/home/janssege/Development/gnucash/builds/master/lib/x86_64-linux-gnu/guile/3.0/site-ccache:/home/janssege/Development/gnucash/builds/master/lib/x86_64-linux-gnu/guile/3.0/site-ccache/gnucash/deprecated:/home/janssege/Development/gnucash/builds/master/lib/x86_64-linux-gnu/guile/3.0/site-ccache/tests
93:  GUILE_WARN_DEPRECATED=detailed
93: Test timeout computed to be: 10000000
93: Passing a string as a type-name to make-record-type is deprecated.
93:   Pass a symbol instead.
93: Backtrace:
93: In ice-9/boot-9.scm:
93:   1736:10  9 (with-exception-handler _ _ #:unwind? _ # _)
93: In unknown file:
93:            8 (apply-smob/0 #<thunk 55f1d69948a0>)
93: In ice-9/boot-9.scm:
93:     718:2  7 (call-with-prompt _ _ #<procedure default-prompt-handle?>)
93: In ice-9/eval.scm:
93:     619:8  6 (_ #(#(#<directory (guile-user) 55f1d6a66f00>)))
93: In ice-9/command-line.scm:
93:    185:18  5 (_ #<input: string 55f1d6a60770>)
93: In unknown file:
93:            4 (eval (exit (run-test)) #<directory (guile-user) 55f1d6?>)
93: In ice-9/eval.scm:
93:    191:35  3 (_ #f)
93: In test-invoice.scm:
93:      43:2  2 (run-test)
93:      75:4  1 (inv-tests invoice)
93: In test-report-extras.scm:
93:      90:8  0 (_ parser-error #<input: string 55f1d6c1bf50> "XMLNS [?" ?)
93: 
93: test-report-extras.scm:90:8: Throw to key `parser-error' with args `((#<input: string 55f1d6c1bf50> "XMLNS [4] for '" #\< "'"))'.
93: 
93: 
93: **** starting tests for variant invoice ****
93: 
93: d0.0279 t0.028: ('not-record: Invoice #)
93: d0.0001 t0.028: ('not-record: #f)
93: d0.0001 t0.028: ('not-record: Date:)
93: d0.0001 t0.028: ('not-record: 09/01/80)
93: d0.0001 t0.028: ('not-record: Due Date:)
93: d0.0001 t0.028: ('not-record: 09/01/80)
93: d0.0000 t0.028: ('not-record: Reference:)
93: d0.0002 t0.029: ('not-record: cust-1-name)
93: d0.0002 t0.029: ('not-record: #f)
93: d0.0001 t0.029: ('not-record: 06/25/20)
93: d0.0002 t0.029: ('not-record: Date)
93: d0.0001 t0.029: ('not-record: Description)
93: d0.0002 t0.029: ('not-record: Action)
93: d0.0000 t0.029: ('not-record: Quantity)
93: d0.0000 t0.029: ('not-record: Unit Price)
93: d0.0000 t0.029: ('not-record: Discount)
93: d0.0001 t0.030: ('not-record: Taxable)
93: d0.0000 t0.030: ('not-record: Tax Amount)
93: d0.0003 t0.030: ('not-record: Total)
93: d0.0000 t0.030: ('not-record: 06/25/20)
93: d0.0000 t0.030: ('not-record: entry-1-desc)
93: d0.0000 t0.030: ('not-record: entry-1-action)
93: d0.0001 t0.030: ('not-record: 2)
93: d0.0005 t0.031: ('not-record: 0.00 %)
93: d0.0001 t0.031: ('not-record: T)
93: d0.0001 t0.031: ('not-record: Net Price)
93: d0.0001 t0.031: ('not-record: Total Price)
93: d0.0001 t0.031: ('not-record: Amount Due)
93: *** XML error: test-invoice inv-1 simple entry: parser-error / (#<input: string 55f1d6c1bf50> XMLNS [4] for ' < ')
93: General / Invoice Number: #<swig-pointer GncInvoice * 55f1d7071070>
93: Display Columns / Tax Amount: Enabled
93: Display / Job Details: Enabled
93: Display / Use Detailed Tax Summary: Enabled
93: Display / Invoice Notes: Enabled
1/1 Test #93: test-invoice .....................***Failed    0.21 sec
Passing a string as a type-name to make-record-type is deprecated.
  Pass a symbol instead.
Backtrace:
In ice-9/boot-9.scm:
  1736:10  9 (with-exception-handler _ _ #:unwind? _ # _)
In unknown file:
           8 (apply-smob/0 #<thunk 55f1d69948a0>)
In ice-9/boot-9.scm:
    718:2  7 (call-with-prompt _ _ #<procedure default-prompt-handle?>)
In ice-9/eval.scm:
    619:8  6 (_ #(#(#<directory (guile-user) 55f1d6a66f00>)))
In ice-9/command-line.scm:
   185:18  5 (_ #<input: string 55f1d6a60770>)
In unknown file:
           4 (eval (exit (run-test)) #<directory (guile-user) 55f1d6?>)
In ice-9/eval.scm:
   191:35  3 (_ #f)
In test-invoice.scm:
     43:2  2 (run-test)
     75:4  1 (inv-tests invoice)
In test-report-extras.scm:
     90:8  0 (_ parser-error #<input: string 55f1d6c1bf50> "XMLNS [?" ?)

test-report-extras.scm:90:8: Throw to key `parser-error' with args `((#<input: string 55f1d6c1bf50> "XMLNS [4] for '" #\< "'"))'.


**** starting tests for variant invoice ****

d0.0279 t0.028: ('not-record: Invoice #)
d0.0001 t0.028: ('not-record: #f)
d0.0001 t0.028: ('not-record: Date:)
d0.0001 t0.028: ('not-record: 09/01/80)
d0.0001 t0.028: ('not-record: Due Date:)
d0.0001 t0.028: ('not-record: 09/01/80)
d0.0000 t0.028: ('not-record: Reference:)
d0.0002 t0.029: ('not-record: cust-1-name)
d0.0002 t0.029: ('not-record: #f)
d0.0001 t0.029: ('not-record: 06/25/20)
d0.0002 t0.029: ('not-record: Date)
d0.0001 t0.029: ('not-record: Description)
d0.0002 t0.029: ('not-record: Action)
d0.0000 t0.029: ('not-record: Quantity)
d0.0000 t0.029: ('not-record: Unit Price)
d0.0000 t0.029: ('not-record: Discount)
d0.0001 t0.030: ('not-record: Taxable)
d0.0000 t0.030: ('not-record: Tax Amount)
d0.0003 t0.030: ('not-record: Total)
d0.0000 t0.030: ('not-record: 06/25/20)
d0.0000 t0.030: ('not-record: entry-1-desc)
d0.0000 t0.030: ('not-record: entry-1-action)
d0.0001 t0.030: ('not-record: 2)
d0.0005 t0.031: ('not-record: 0.00 %)
d0.0001 t0.031: ('not-record: T)
d0.0001 t0.031: ('not-record: Net Price)
d0.0001 t0.031: ('not-record: Total Price)
d0.0001 t0.031: ('not-record: Amount Due)
*** XML error: test-invoice inv-1 simple entry: parser-error / (#<input: string 55f1d6c1bf50> XMLNS [4] for ' < ')
General / Invoice Number: #<swig-pointer GncInvoice * 55f1d7071070>
Display Columns / Tax Amount: Enabled
Display / Job Details: Enabled
Display / Use Detailed Tax Summary: Enabled
Display / Invoice Notes: Enabled


0% tests passed, 1 tests failed out of 1

Total Test time (real) =   0.22 sec

The following tests FAILED:
	 93 - test-invoice (Failed)
Errors while running CTest
```

**Comment by christopherlam** [comment]
To simplify this PR, I'll change all _ to G_ in the code soon so we can focus on getting the upgrade done.

**Comment by christopherlam** [comment]
Please repush guile-3 changes...

**Comment by christopherlam** [comment]
Moreover the PR must now be edited to target maint.

**Comment by jralls** [comment]
In light of #754 ISTM this should be closed and a new PR made to deal with the other guile3 incompatibilities.

Deprecated function removals should be in their own commit. (```SCM_LISTN``` was deprecated in **2003**! Talk about technical debt, both ours and guile's!) Let's not stop at the deprecated forms removed in guile 3: Everything deprecated in guile 2 and earlier should get replaced.

**Comment by christopherlam** [comment]
Thanks, we managed to add support for guile-3 by doing piecemeal changes in the `maint` tree. Please confirm!

**Pull request closed without merging**